internal/generator/config/curve.go

```diff
@@ -69,3 +94,178 @@ func newFieldInfo(modulus string) Field {
 	F.Modulus = func() *big.Int { return new(big.Int).Set(&bModulus) }
 	return F
 }
+
+// twoFactor returns the greatest m such that 2ᵐ divides x-1
+func twoFactor(x *big.Int) int {
+	if x.BitLen() <= 1 {
+		return -1 //substitute for ∞
+	}
+	if x.Bit(0) == 0 {
+		return 0 //rightmost bit of x-1 is 1
+	}
+	var m int
+	for m = 1; x.Bit(m) == 0; m++ {
+	}
+	return m
+}
+
+func NewHashSuiteInfo(fieldModulus *big.Int, g *Point, name string, suite *HashSuite) HashSuiteInfo {
+
+	fieldSize := pow(fieldModulus, g.CoordExtDegree)
+	fieldSizeMod256 := uint8(fieldSize.Bits()[0])
+
+	Z := int64(suite.Z)
+	var c []big.Int
+
+	//TODO: Works only for fp
+	if fieldSizeMod256%4 == 3 {
+		c = make([]big.Int, 2)
+		c[0].Rsh(fieldSize, 2)
+
+		c[1].SetInt64(-Z)
+		c[1].ModSqrt(&c[1], fieldModulus)
+		field.IntToMont(&c[1], fieldModulus)
+
+	} else if fieldSizeMod256%8 == 5 {
+		c = make([]big.Int, 3)
+		c[0].Rsh(fieldSize, 3)
+
+		c[1].SetInt64(-1)
+		c[1].ModSqrt(&c[1], fieldModulus)
+
+		c[2].ModInverse(&c[1], fieldModulus)
+		c[2].Mul(&c[2], big.NewInt(Z))
+
+		c[2].ModSqrt(&c[2], fieldModulus)
+
+		field.IntToMont(&c[1], fieldModulus)
+		field.IntToMont(&c[2], fieldModulus)
+	} else if fieldSizeMod256%8 == 1 {
+		ONE := big.NewInt(1)
+		c = make([]big.Int, 7)
+		c1 := twoFactor(fieldSize)
+		c[0].SetInt64(int64(c1))
+		var twoPowC1 big.Int
+		twoPowC1.Lsh(ONE, uint(c1))
+		c[1].Rsh(fieldSize, uint(c1))
+		c[2].Rsh(&c[1], 1)
+		c[3].Sub(&twoPowC1, ONE)
+		c[4].Rsh(&twoPowC1, 1)
+		powMod(&c[5], big.NewInt(Z), &c[1], fieldModulus)
+		var c7Pow big.Int
+		c7Pow.Add(&c[1], ONE)
+		c7Pow.Rsh(&c7Pow, 1)
+		powMod(&c[6], big.NewInt(Z), &c7Pow, fieldModulus)
+
+		field.IntToMont(&c[5], fieldModulus)
+		field.IntToMont(&c[6], fieldModulus)
+
+	} else {
+		panic("this is logically impossible")
+	}
+
+	return HashSuiteInfo{
+		A:                field.HexToMont(suite.AHex, fieldModulus),
+		B:                field.HexToMont(suite.BHex, fieldModulus),
+		Z:                suite.Z,
+		CoordType:        g.CoordType,
+		CofactorCleaning: g.CofactorCleaning,
+		Name:             name,
+		Isogeny:          newIsogenousCurveInfoOptional(fieldModulus, suite.Isogeny),
+		FieldSizeMod256:  fieldSizeMod256,
+		SqrtRatioParams:  c,
+	}
+}
+
+func powMod(res *big.Int, x *big.Int, pow *big.Int, mod *big.Int) *big.Int {
+	res.SetInt64(1)
+
+	for i := pow.BitLen() - 1; ; {
+
+		if pow.Bit(i) == 1 {
+			res.Mul(res, x)
+		}
+
+		if i == 0 {
+			break
+		}
+		i--
+
+		res.Mul(res, res).Mod(res, mod)
+	}
+
+	res.Mod(res, mod)
+	return res
+}
+
+func pow(p *big.Int, pow uint8) *big.Int {
```

Comment:
why not use big.Int Exp method here? for powMod I understand big.Int Exp followed by a mod is going to be terribly inefficient. but simple pow?

Reply:
I got the impression from the API that only modular exponentiation is supported. It seems like setting the modulus to nil gives integer exp.